### PR TITLE
feat: [IDP-1430]: Status info v2 API

### PR DESCRIPTION
**Pull request by vigneswara-propelo**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/55106)

### CODE REVIEW

**Review by vikyathharekal — COMMENTED** [review]
> Reviewed 6 of 7 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/55106)* status: 7 file(s) to review. 0 of 2 approvals obtained. 

___
*[`idp-service/contracts/openapi/v1/openapi.yaml` line 3599 at r1](https://reviewable.io/reviews/harness/harness-core/55106#-NidyJEO5QN83yjyXuUt:-NidyJEO5QN83yjyXuUu:b-ixtne9) ([raw file](https://github.com/harness/harness-core/blob/20065af569753369f0c09e8c5cc6238c114508ad/idp-service/contracts/openapi/v1/openapi.yaml#L3599)):*
> ```YAML
>           - infra
>           - onboarding
>           - infra_onboarding
> ```

If someone hits the old API with the "infra_onboarding" type, it would return NOT_FOUND. Would that be fine?



_Code quote:_
```YAML
- infra_onboarding
```
___
*[`idp-service/src/main/java/io/harness/idp/status/service/StatusInfoServiceImpl.java` line 49 at r1](https://reviewable.io/reviews/harness/harness-core/55106#-Nidxks2BhwDAOCz1wPq:-Nidxks2BhwDAOCz1wPr:bn8ptxk) ([raw file](https://github.com/harness/harness-core/blob/20065af569753369f0c09e8c5cc6238c114508ad/idp-service/src/main/java/io/harness/idp/status/service/StatusInfoServiceImpl.java#L49)):*
> ```Java
>   public StatusInfoV2 findByAccountIdentifierAndTypeV2(String accountIdentifier, String type) {
>     StatusInfoV2 statusInfoV2 = new StatusInfoV2();
>     switch (type) {
> ```

How about this?
```
switch (StatusType.valueOf(type)) {  
case INFRA:
...
case ONBOARDING:
...
```
This would avoid the strings used below
___
*[`idp-service/src/main/java/io/harness/idp/status/service/StatusInfoServiceImpl.java` line 69 at r1](https://reviewable.io/reviews/harness/harness-core/55106#-NidvrV0CLq8jritgWnv:-NidvrV0CLq8jritgWnw:bbhe1on) ([raw file](https://github.com/harness/harness-core/blob/20065af569753369f0c09e8c5cc6238c114508ad/idp-service/src/main/java/io/harness/idp/status/service/StatusInfoServiceImpl.java#L69)):*
> ```Java
> 
>   private StatusInfo getStatusInfoForInfra(String accountIdentifier) {
>     return healthCheck.getCurrentStatus(accountIdentifier).get();
> ```

This would throw an exception if not present. Should be fine as we catch the exception in the resource layer and return 500. 
Is this intended? or do we need to check isPresent() and add null or "Unkown" to the StatusInfoV2?



_Code quote:_
```Java
.get()
```

File: idp-service/contracts/openapi/v1/openapi.yaml [hunk 1]
```diff
@@ -276,6 +276,24 @@ paths:
       requestBody:
         $ref: '#/components/requestBodies/StatusInfoRequest'
       x-internal: true
+  /v2/status-info/{type}:
```
Comment (sathish-soundarapandian) [inline idp-service/contracts/openapi/v1/openapi.yaml:279]: We can mark the V1 API as deprecated. Not sure what is the right way to deprecate earlier version in favour of newer version.

**Review by vigneswara-propelo — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/55106)* status: 7 file(s) to review. 4. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`idp-service/contracts/openapi/v1/openapi.yaml` line 279 at r1](https://reviewable.io/reviews/harness/harness-core/55106#gh-1384924621:-Nie4pWbEWDXkcNyAf6t:b-j2j30j) ([raw file](https://github.com/harness/harness-core/blob/20065af569753369f0c09e8c5cc6238c114508ad/idp-service/contracts/openapi/v1/openapi.yaml#L279)):*
<details><summary><i>Previously, sathish-soundarapandian (Sathish) wrote…</i></summary><blockquote>

We can mark the V1 API as deprecated. Not sure what is the right way to deprecate earlier version in favour of newer version.
</blockquote></details>

sure, we can mark it deprecated once UI starts using v2 API



_Code snippet:_
```YAML
deprecated: true
```
___
*[`idp-service/contracts/openapi/v1/openapi.yaml` line 3599 at r1](https://reviewable.io/reviews/harness/harness-core/55106#-NidyJEO5QN83yjyXuUt:-Nie57DXAn6WeS3EVknm:b-a0gwgy) ([raw file](https://github.com/harness/harness-core/blob/20065af569753369f0c09e8c5cc6238c114508ad/idp-service/contracts/openapi/v1/openapi.yaml#L3599)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

If someone hits the old API with the "infra_onboarding" type, it would return NOT_FOUND. Would that be fine?
</blockquote></details>

yeah, that should be fine. Anyways, UI will not be using infra_onboarding in the old API
___
*[`idp-service/src/main/java/io/harness/idp/status/service/StatusInfoServiceImpl.java` line 49 at r1](https://reviewable.io/reviews/harness/harness-core/55106#-Nidxks2BhwDAOCz1wPq:-Nie5etd-p3FwZkIElOR:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/20065af569753369f0c09e8c5cc6238c114508ad/idp-service/src/main/java/io/harness/idp/status/service/StatusInfoServiceImpl.java#L49)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

How about this?
```
switch (StatusType.valueOf(type)) {  
case INFRA:
...
case ONBOARDING:
...
```
This would avoid the strings used below
</blockquote></details>

Done.
___
*[`idp-service/src/main/java/io/harness/idp/status/service/StatusInfoServiceImpl.java` line 69 at r1](https://reviewable.io/reviews/harness/harness-core/55106#-NidvrV0CLq8jritgWnv:-NieAL7q3F4eUVrbMwjb:b-1ltvdc) ([raw file](https://github.com/harness/harness-core/blob/20065af569753369f0c09e8c5cc6238c114508ad/idp-service/src/main/java/io/harness/idp/status/service/StatusInfoServiceImpl.java#L69)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

This would throw an exception if not present. Should be fine as we catch the exception in the resource layer and return 500. 
Is this intended? or do we need to check isPresent() and add null or "Unkown" to the StatusInfoV2?
</blockquote></details>

I think we should be good with this as we construct StatusInfo in all the scenarios inside HealthCheck and also we first initialise it

**Review by vikyathharekal — APPROVED** [review]
> Reviewed 1 of 7 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/55106)* status: 7 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by sathish-soundarapandian — APPROVED** [review]

**Review by mankrit-singh — APPROVED** [review]
Originally, the plan was to support api versioning where the path itself would not change, and users can navigate to different versions with Version Headers.

But since this item has since become deprioritized and there is no ETA on this, we can continue with using "v2" for now in the path if there is no alternative.

Whenever API versioning is prioritized we can then switch back to v1 for consistency and mark v2 as deprecated.

**Pull request merged**